linera-core/src/client/mod.rs

```diff
@@ -462,17 +465,17 @@ impl<Env: Environment> Client<Env> {
             .await?;
         if let Some(blob) = blob {
             // We have the blob - return it.
-            return Ok(blob);
+            return Ok(bcs::from_bytes(blob.bytes())?);
         };
         // Recover history from the current validators, according to the admin chain.
-        // TODO(#2351): make sure that the blob is legitimately created!
+        self.synchronize_chain_state(self.admin_id).await?;
```

Comment:
And again we're synchronizing the admin chain - is this one of the reasons you want to move everything to a single place so that we have a better control over what has been done or what needs to be done?

Reply:
Yes, exactly!
I don't know how to get around these otherwise.

E.g. we shouldn't do this at all if we're in "connected" mode, i.e. a chain listener is running. I'm thinking about writing a design doc for a "local node synchronizer" that does all this properly. I'd rather not try to add more ad-hoc case distinctions here for now.

Reply:
Definitely.